### PR TITLE
Add multi select component to ha-form

### PR DESCRIPTION
**Pull request by bramkragten**
image

### CODE REVIEW

**Comment by Kane610** [comment]
Awesome 🎉

File: src/components/ha-form/ha-form-multi_select.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  query,
+} from "lit-element";
+import {
+  HaFormElement,
+  HaFormMultiSelectSchema,
+  HaFormMultiSelectData,
+} from "./ha-form";
+import { fireEvent } from "../../common/dom/fire_event";
+
+import "@polymer/paper-dropdown-menu/paper-dropdown-menu";
```
Comment (iantrich) [inline src/components/ha-form/ha-form-multi_select.ts]: external imports before internal

File: src/components/ha-form/ha-form-multi_select.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import "@polymer/paper-checkbox/paper-checkbox";
+
+@customElement("ha-form-multi_select")
+export class HaFormMultiSelect extends LitElement implements HaFormElement {
+  @property() public schema!: HaFormMultiSelectSchema;
+  @property() public data!: HaFormMultiSelectData;
+  @property() public label!: string;
+  @property() public suffix!: string;
+  @property() private _init = false;
+  @query("paper-dropdown-menu") private _input?: HTMLElement;
```
Comment (iantrich) [inline src/components/ha-form/ha-form-multi_select.ts]: ooo didn't know about this decorator. cool

File: src/components/ha-form/ha-form-multi_select.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  public focus() {
```
Comment (iantrich) [inline src/components/ha-form/ha-form-multi_select.ts]:
```suggestion
  public focus(): void {
```

File: src/components/ha-form/ha-form-multi_select.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+              ></paper-checkbox>
+              ${this._optionLabel(item)}
+            </paper-icon-item>
+          `;
+        })}
+      </paper-listbox>
+    `;
+  }
+
+  private _optionValue(item) {
```
Comment (iantrich) [inline src/components/ha-form/ha-form-multi_select.ts]:
```suggestion
  private _optionValue(item: string | string[]): string | string[] {
```

File: src/components/ha-form/ha-form-multi_select.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    return Array.isArray(item) ? item[0] : item;
+  }
+
+  private _optionLabel(item) {
```
Comment (iantrich) [inline src/components/ha-form/ha-form-multi_select.ts]:
```suggestion
  private _optionLabel(item: string | string[]): string | string[] {
```

File: src/components/ha-form/ha-form-multi_select.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    return Array.isArray(item) ? item[1] : item;
+  }
+
+  private _valueChanged(ev: CustomEvent) {
```
Comment (iantrich) [inline src/components/ha-form/ha-form-multi_select.ts]:
```suggestion
  private _valueChanged(ev: CustomEvent): void {
```

File: src/components/ha-form/ha-form.ts [hunk 7]
```diff
@@ -45,6 +47,11 @@ export interface HaFormSelectSchema extends HaFormBaseSchema {
   options?: string[];
 }
 
+export interface HaFormMultiSelectSchema extends HaFormBaseSchema {
+  type: "multi_select";
```
Comment (balloob) [inline src/components/ha-form/ha-form.ts]: I wonder if this should be it's own interface or that the select schema should get a `multi=true` option passed in via the description ? 

```python
vol.Schema({
  vol.Required("something", description={"multi": True}): vol.In(options_dict)
})
```

**Comment by iantrich** [comment]
is there a backend PR for this to go against?

**Comment by Kane610** [comment]
@iantrich no not yet

File: src/components/ha-form/ha-form.ts (hunk 7 above)
Comment (balloob) [reply src/components/ha-form/ha-form.ts]: I think that this comment is how we can get it into the backend today.

File: src/components/ha-form/ha-form.ts [hunk 8]
```diff
@@ -45,6 +47,11 @@ export interface HaFormSelectSchema extends HaFormBaseSchema {
...
+  options?: string[];
```
Comment (balloob) [inline src/components/ha-form/ha-form.ts]: Backend support has been added https://github.com/home-assistant/home-assistant/pull/31798

This has become dictionary of key=>label: `{ [key: string]: string }

**Comment by Kane610** [comment]
Woo-hoo 🎉

**Pull request merged**